### PR TITLE
Aruha-664 new command

### PR DESCRIPTION
**Pull request by lmontrieux**
This PR adds two options to the `rebalance` command

* the first option, `--empty-brokers`, takes a comma-separated list of brokers from which all partitions should be removed. The partitions in questions will be relocated to other brokers
* the second option, `--exclude-topics`, takes a comma-separated list of topics that should be excluded from rebalance, i.e. they will not be moved

### CODE REVIEW

File: bubuku/cli.py [hunk 1]
```diff
@@ -204,5 +208,65 @@ def show_stats():
...
+                broker_ids = tp_to_brokers[candidate]
+                print('old allocation for {}: {}'.format(candidate, broker_ids))
+                length = len(broker_ids)
+                del broker_ids[not_loaded]
+                broker_ids = [not_loaded] + broker_ids
+                if length != len(broker_ids):
+                    raise Exception('broker ids length changed')
+                print('New allocation for candidate {}: {}'.format(candidate, broker_ids))
+                # ex.reallocate_partition(candidate[0], candidate[1], broker_ids)
+                return
```
Comment (antban) [inline bubuku/cli.py]: What is that?

File: bubuku/features/rebalance/change.py [hunk 2]
```diff
@@ -134,11 +135,11 @@ def run(self, current_actions) -> bool:
...
-        new_broker_ids = sorted(int(id_) for id_ in self.zk.get_broker_ids())
-        if new_broker_ids != self.broker_ids:
-            _LOG.warning("Rebalance stopped because of broker list change from {} to {}".format(
-                self.broker_ids, new_broker_ids))
-            return False
+        # new_broker_ids = sorted(int(id_) for id_ in self.zk.get_broker_ids())
+        # if new_broker_ids != self.broker_ids:
+        #     _LOG.warning("Rebalance stopped because of broker list change from {} to {}".format(
+        #         self.broker_ids, new_broker_ids))
+        #     return False
```
Comment (antban) [inline bubuku/features/rebalance/change.py]: No, this should be done in different manner. 
The purpose of this check is to stop rebalance when one of nodes goes down or new node created (For example one decided to restart node). It's ok to remove it when the process is manually started, but with automatic process one should still take care of this situation

File: bubuku/cli.py (hunk 1 above)
Comment (lmontrieux) [inline bubuku/cli.py]: Your code, as I branched from the 'migrate-data' branch (but removed the handmade test, as I don't think it should be merged to master in that state)

File: bubuku/features/rebalance/change.py [hunk 3]
```diff
@@ -155,6 +156,28 @@ def run(self, current_actions) -> bool:
...
+
+        def _get_size(v):
+            for ds in disk_stats.values():
+                size = ds['topics'].get(v[0], {}).get(str(v[1]))
+                if size is not None and size > 0:
+                    return size
+            return 0
+
+        actions_sorted_by_size = sorted(change.action_queue, key=_get_size)
+        # actions_sorted_by_size = change.action_queue
```
Comment (antban) [inline bubuku/features/rebalance/change.py]: Well, there is actually problem here. It will sort only changes, not the source data. IT works when it's a handmade process, but with real process one should put more effort to do it better, or at least add comment about that in code

**Comment by rcillo** [comment]
@lmontrieux I could not find a test that combines both parameters. It would be interesting to do so, because in reality, it would be useful to empty broker with exception of `__consumer_offsets`, which I would not like to move.

**Comment by lmontrieux** [comment]
@rcillo ad1399c

File: bubuku/features/rebalance/change.py (hunk 2 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/change.py]: If I understand correctly, it's ok to keep it even when the process is manually started?

File: bubuku/cli.py (hunk 1 above)
Comment (lmontrieux) [inline bubuku/cli.py]: I'm going to remove this entirely, if we want to merge it from 'migrate-data', it should be in a different PR (and with tests :) )

File: bubuku/features/rebalance/change.py (hunk 3 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/change.py]: Actually, since I remove the handmade test, this function is not used anywhere, so I will also remove it

Comment (lmontrieux) [inline bubuku/features/rebalance/change.py]: d907146

File: bubuku/features/rebalance/change.py (hunk 2 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/change.py]: Now I understand why this was commented out. Fixed in dfac651

File: bubuku/cli.py (hunk 1 above)
Comment (lmontrieux) [inline bubuku/cli.py]: 2754d8f

File: bubuku/cli.py [hunk 4]
```diff
@@ -72,11 +72,15 @@ def restart_broker(broker: str):
...
+@click.option('--empty_brokers', type=click.STRING,
+              help="Comma-separated list of brokers to empty. All partitions will be moved to other brokers")
+@click.option('--exclude_topics', type=click.STRING, help="Comma-separated list of topics to exclude from rebalance")
+def rebalance_partitions(broker: str, empty_brokers: str, exclude_topics: str):
     config, env_provider = __prepare_configs()
     with load_exhibitor_proxy(env_provider.get_address_provider(), config.zk_prefix) as zookeeper:
         broker_id = __get_opt_broker_id(broker, config, zookeeper, env_provider) if broker else None
-        RemoteCommandExecutorCheck.register_rebalance(zookeeper, broker_id)
+        RemoteCommandExecutorCheck.register_rebalance(zookeeper, broker_id, empty_brokers.split(','),
+                                                      exclude_topics.split(','))
```
Comment (antban) [inline bubuku/cli.py]: empty_brokers and exclude_topics may be `None`, so method will fail

File: bubuku/features/rebalance/change.py [hunk 5]
```diff
@@ -135,9 +137,9 @@ def run(self, current_actions) -> bool:
         if self.zk.is_rebalancing():
             return True
         new_broker_ids = sorted(int(id_) for id_ in self.zk.get_broker_ids())
-        if new_broker_ids != self.broker_ids:
+        if new_broker_ids != (sorted(int(id_) for id_ in self.all_broker_ids)):
```
Comment (antban) [inline bubuku/features/rebalance/change.py]: Why to sort several times? it is already sorted

File: bubuku/features/rebalance/check.py [hunk 6]
```diff
@@ -21,7 +21,7 @@ def check(self):
             return None
         _LOG.info("Rebalance on start, triggering rebalance")
         self.executed = True
-        return OptimizedRebalanceChange(self.zk, self.zk.get_broker_ids())
+        return OptimizedRebalanceChange(self.zk, self.zk.get_broker_ids(), [], False)
```
Comment (antban) [inline bubuku/features/rebalance/check.py]: False is not a list

File: bubuku/features/rebalance/check.py [hunk 7]
```diff
@@ -41,7 +41,7 @@ def check(self):
         if not new_list == self.old_broker_list:
             _LOG.info('Broker list changed from {} to {}, triggering rebalance'.format(self.old_broker_list, new_list))
             self.old_broker_list = new_list
-            return OptimizedRebalanceChange(self.zk, new_list)
+            return OptimizedRebalanceChange(self.zk, new_list, [], False)
```
Comment (antban) [inline bubuku/features/rebalance/check.py]: The same here

File: bubuku/cli.py (hunk 4 above)
Comment (lmontrieux) [inline bubuku/cli.py]: 77549f6 and 4cf5ae6

File: bubuku/features/rebalance/change.py (hunk 5 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/change.py]: 4e6e4bf

File: bubuku/features/rebalance/check.py (hunk 7 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/check.py]: b302a21

File: bubuku/features/rebalance/check.py (hunk 6 above)
Comment (lmontrieux) [inline bubuku/features/rebalance/check.py]: b302a21

**Comment by antban** [comment]
:+1:

**Comment by lmontrieux** [comment]
@antban could you please approve again?

**Comment by antban** [comment]
:+1:

**Comment by antban** [comment]
:+1:

**Comment by lmontrieux** [comment]
👍
